File: Python/2015/Wait Time.py
# J4

commands = int(input())
lines = []
people = []
temptimes = {}
finaltimes = {}
time = 0

for i in range(commands):
    x = input().split()
    x[1] = int(x[1])
    lines.append(x)

for line in lines:
    if line[0] == 'R':
        temptimes[line[1]] = time
        if line[1] not in people:
            people.append(line[1])
        time     += 1
    elif line[0] == 'W':
        time += (line[1] - 1)
    elif line[0] == 'S':
        if finaltimes.get(line[1]) is None:
            finaltimes[line[1]] = (time - temptimes.get(line[1]))
        else:
            finaltimes[line[1]] += (time - temptimes.get(line[1]))
        del temptimes[line[1]]
        time += 1

people.sort()

for person in people:
    if temptimes.get(person) is None:
        print(person, finaltimes.get(person))
    else:
        print(person, -1)

File: C++/2015/Wait Time.cpp
// J4

#include <iostream>
#include <map>
#include <vector>
#include <tuple>
#include <algorithm>

using namespace std;

tuple<char, int> split(string line)
{
    char command;
    int length = line.length(), num;
    if (length == 3)
    {
        command = line.at(0);
        num = (int)line.at(2) - 48;
    }
    else
    {
        command = line.at(0);
        num = stoi(line.substr(2));
    }

    return make_tuple(command, num);
}

int main()
{
    int numcommands, time = 0;
    cin >> numcommands;

    vector<tuple<char, int>> commands;
    vector<int> people;
    map<int, int> temptimes, finaltimes;
    cin.ignore();
    for (int i = 0; i < numcommands; i++)
    {
        string input;
        getline(cin, input);
        commands.push_back(split(input));
    }

    for (tuple<char, int> command : commands)
    {
        char action = get<0>(command);
        int num = get<1>(command);

        if (action == 'R')
        {
            temptimes[num] = time;
            if (find(people.begin(), people.end(), num) == people.end())
            {
                people.push_back(num);
            }
            time += 1;
        }
        else if (action == 'W')
        {
            time += (num - 1);
        }
        else if (action == 'S')
        {
            if (finaltimes.find(num) == finaltimes.end())
            {
                finaltimes[num] = (time - temptimes[num]);
            }
            else
            {
                finaltimes[num] += (time - temptimes[num]);
            }
            temptimes.erase(num);
            time += 1;
        }
    }

    sort(people.begin(), people.end());

    for (int i = 0; i < people.size(); i++)
    {
        int person = people[i];
        if (temptimes.find(person) == temptimes.end())
        {
            cout << to_string(person) << " " << to_string(finaltimes[person]) << endl;
        }
        else
        {
            cout << to_string(person) << " " << to_string(-1) << endl;
        }
    }
}